src/query/sql/src/planner/optimizer/heuristic/prewhere_optimization.rs

```diff
@@ -11,133 +11,133 @@
-                for arg in func.arguments.iter() {
-                    Self::collect_columns_impl(arg, columns);
-                }
-            }
-            Scalar::CastExpr(cast) => {
-                Self::collect_columns_impl(cast.argument.as_ref(), columns);
-            }
-            // 1. ConstantExpr is not collected.
-            // 2. SubqueryExpr and AggregateFunction will not appear in Filter-LogicalGet
-            _ => {}
-        }
-    }
-
-    // analyze if the expression can be moved to prewhere
-    fn collect_columns(expr: &Scalar) -> ColumnSet {
-        let mut columns = ColumnSet::new();
-        // columns in subqueries are not considered
-        Self::collect_columns_impl(expr, &mut columns);
-
-        columns
-    }
-
-    pub fn prewhere_optimize(&self, s_expr: SExpr) -> Result<SExpr> {
-        let rel_op = s_expr.plan();
-        if s_expr.match_pattern(&self.pattern) {
-            let filter: Filter = s_expr.plan().clone().try_into()?;
-            let mut get: LogicalGet = s_expr.child(0)?.plan().clone().try_into()?;
-            get.push_down_predicates = Some(filter.predicates.clone());
-            let metadata = self.metadata.read().clone();
-
-            let table = metadata.table(get.table_index).table();
-            if !table.support_prewhere() {
-                // cannot optimize
-                return Ok(s_expr);
-            }
-
-            let mut prewhere_columns = ColumnSet::new();
-            let mut prewhere_pred = Vec::new();
-
-            // filter.predicates are already splited by AND
-            for pred in filter.predicates.iter() {
-                let columns = Self::collect_columns(pred);
-                prewhere_pred.push(pred.clone());
-                prewhere_columns.extend(&columns);
-            }
-
-            get.prewhere = if prewhere_pred.is_empty() {
-                None
-            } else {
-                if prewhere_columns.is_empty() {
-                    // select a, b from t1 where 's' not in ('x', 'b', 'c', 'd') or to_nullable(null);
-                    // get the smallest column in (a,b). (maybe can get the smallest column in t1)
-                    let used = self.metadata.read().find_smallest_column(
-                        get.columns.iter().copied().collect::<Vec<_>>().as_slice(),
-                    );
-                    prewhere_columns.insert(used);
-                }
-                Some(Prewhere {
-                    output_columns: get.columns.clone(),
-                    prewhere_columns,
-                    predicates: prewhere_pred,
-                })
-            };
-
-            Ok(SExpr::create_leaf(get.into()))
-        } else {
-            let children = s_expr
-                .children()
-                .iter()
-                .map(|expr| self.prewhere_optimize(expr.clone()))
-                .collect::<Result<Vec<_>>>()?;
-            Ok(SExpr::create(rel_op.clone(), children, None, None))
-        }
-    }
-}
+//
```

Comment:
`prewhere` may be useful for other engines like Parquet engine.

We can turn off the prewhere in catalog:

```
fn support_prewhere(&self) -> bool {
        false
}
```

For fusetable.